content/blog/2024/09/usr-merge/index.md

```diff
@@ -0,0 +1,90 @@
+
+### The plan
+
+The process goes like this:
+
+1) eopkg is updated with Usr-Merge support. It will write a marker file to indicate that the system can safely be merged.
+2) Our Usr Merge script runs during early boot and detects the marker file. The script proceeds to upgrade the machine safely, checking every file along the way with each step, only continuing if everything is correct.
+
+   The checking does mean that the script takes a while to run on less powerful systems; our testing in a Virtual Machine simulating really old hardware showed that the script took about 3 minutes to complete. Recent package updates include work to speed up the migration process significantly by essentially self-migrating, but it could still take a while.
+   
+   Because of the time the script could take to run, a message will be shown on screen so the user knows that something is happening. The script is robust enough to not destory your system if a reboot occurs while it is running, but doing a hard reboot does mean that the script has to start over on the next boot.
+   
+   Lastly, a marker file is written that indicates that the system is now Usr Merged.
+3) eopkg will use the second marker file to know that it is safe to switch to a new package repository that is incompatible with non-Usr-Merged systems.
+
+   The switch will be activated when the phased rollout is complete.
+
+Once the migration has rolled out to everyone (more on this later), we can fork our package repository and do an Epoch bump, where we can safely release package updates, knowing that they'll only be going out to Usr Merged systems. While an Epoch bump isn't strictly necessary for the merging process itself, it *is* necessary to safely remove our pre-Usr-Merge compatability symlinks from various packages. Removing the symlinks on a system that has not been merged will likely break the system, as we saw last time, because update order is non-deterministic.
+
+Migration to a new package repository has actually been done before on Solus. Most recently, it was done when the [repositories were moved](https://getsol.us/2018/09/11/package-repo-migration-now-available/) to the infrastructure at the Rochester Institute of Technology.
```

Comment:
Actually most recently it was done after relaunching the infrastructure.

Reply:
https://github.com/getsolus/eopkg/commit/ea2f1638d8c3edbe491e3f9dc9e2fcd6751a39f2